lib/ar_state_machine.rb

```diff
@@ -0,0 +1,272 @@
+          end 
+        end
+
+        self.class.send :define_method, ss do 
+          where(state: ss)  
+        end
+
+        self.class.send :define_method, "not_#{ss}" do 
+          where.not(state: ss)
+        end
+      end
+    end
+
+    # this sets all the state names as a constant on the class so we can do things like Product::ACTIVE 
+    # instead of having string everywhere
+    def const_missing(name)
+      state_name = name.downcase
+      return const_set(name, state_name.to_sym) if self.states.keys.include?(state_name)
+      super(name)
+    end
+
+    def before_transition_to(to, method=nil, rollback_on_failure=true, &block)
+      if to.class != Array
+        to = [to]
+      end
+      to.each do |to_|
+        @before_transitions[to_.to_sym] ||= []
+        if !method.present?
+          method = "_before_transitions_#{to_.to_sym}_#{@before_transitions[to_.to_sym].length}"
+          define_method method, block
+        end
+
+        @before_transitions[to_.to_sym].push({method: method, rollback_on_failure: rollback_on_failure})
+      end
+      true
+    end
+
+    def after_transition_to(to, method=nil, &block)
+      if to.class != Array
+        to = [to]
+      end
+      to.each do |to_|
+        @after_transitions[to_.to_sym] ||= []
+        if !method.present?
+          method = "_after_transition_to_#{to_.to_sym}_#{@after_transitions[to_.to_sym].length}"
+          define_method method, block
+        end
+
+        @after_transitions[to_.to_sym].push({method: method}) # AR doesn't do rollbacks for after_* callbacks
+      end
+      true
+    end
+
+    def process_callbacks(to, model, from, callbacks)
+      callbacks.each do |callback|
+        args = case model.method(callback[:method]).parameters.count
+        when 1
+          [from.to_sym]
+        when 2
+          [from.to_sym, to.to_sym]
+        else
+          []
+        end
+
+        if model.send(callback[:method], *args) == false
+          
+          # cancel others if any returned false
+
+          if model.state == to # revert the state if AR won't and it hasn't been changed already
+            model.state = from 
+            if model.respond_to?("#{to}_at=")
+              model.send("#{to}_at=", nil)
+            end
+          end
+
+          if callback[:rollback_on_failure] == false
+            break # just exit THIS chain; not the whole rails callback chain
+          else
+            # rollback changes / cancel subsequent other callbacks
+            return false
+          end
+        end
+      end 
+      true
+    end
+
+    def run_before_transition_callbacks(to, model, from)
+      callbacks = @before_transitions[to.to_sym]
+      return process_callbacks(to, model, from, callbacks) if callbacks
+    end
+
+    def run_after_transition_callbacks(to, model, from)
+      callbacks = @after_transitions[to.to_sym]
+      return process_callbacks(to, model, from, callbacks) if callbacks
+    end
+  end
+end
```

Comment:
I think here you want ```ActiveRecord::Base.extend(StateMachine::ActiveRecordExtensions)``` so that ```self.state_machine``` is available in all AR models